modules/dop/services/workbench/workbench.go

```diff
@@ -0,0 +1,194 @@
+// display undone issue, issues order by priority
+var (
+	issueUnspecified = "unspecified"
+	issueExpired     = "expired"
+	issueOneDay      = "oneDay"
+	issueTomorrow    = "tomorrow"
+	issueSevenDay    = "sevenDay"
+	issueThirtyDay   = "thirtyDay"
+	issueFeature     = "feature"
+	expireDays       = []string{issueUnspecified, issueExpired, issueOneDay, issueTomorrow, issueSevenDay, issueThirtyDay, issueFeature}
+	StateBelongs     = []apistructs.IssueStateBelong{
+		apistructs.IssueStateBelongReopen,
+		apistructs.IssueStateBelongWontfix,
+		apistructs.IssueStateBelongResloved,
+		apistructs.IssueStateBelongWorking,
+		apistructs.IssueStateBelongOpen,
+	}
+	IssuePriorities = []apistructs.IssuePriority{
+		apistructs.IssuePriorityUrgent,
+		apistructs.IssuePriorityHigh,
+		apistructs.IssuePriorityNormal,
+		apistructs.IssuePriorityLow,
+	}
+	IssueTypes = []apistructs.IssueType{
+		apistructs.IssueTypeRequirement,
+		apistructs.IssueTypeBug,
+		apistructs.IssueTypeTask,
+	}
+)
+
+type Workbench struct {
+	bdl      *bundle.Bundle
+	issueSvc *issue.Issue
+}
+
+type Option func(*Workbench)
+
+func New(options ...Option) *Workbench {
+	is := &Workbench{}
+	for _, op := range options {
+		op(is)
+	}
+	return is
+}
+
+// WithIssue set issue service
+func WithIssue(i *issue.Issue) Option {
+	return func(w *Workbench) {
+		w.issueSvc = i
+	}
+}
+
+// WithBundle set bundle
+func WithBundle(bdl *bundle.Bundle) Option {
+	return func(w *Workbench) {
+		w.bdl = bdl
+	}
+}
+
+// GetUndoneProjectItem query issue list and use SetSpecialIssueNum func set issue num
+func (w *Workbench) GetUndoneProjectItem(userID string, issueSize int, pro apistructs.ProjectDTO) (*apistructs.WorkbenchProjectItem, error) {
+	var issueItem apistructs.WorkbenchProjectItem
+	issueItem.IssueList = make([]apistructs.Issue, 0)
+	issueReq := apistructs.IssuePagingRequest{
+		OrgID:    int64(pro.OrgID),
+		PageNo:   1,
+		PageSize: uint64(issueSize),
+		IssueListRequest: apistructs.IssueListRequest{
+			ProjectID:    uint64(pro.ID),
+			StateBelongs: StateBelongs,
+			Assignees:    []string{userID},
+			External:     true,
+			OrderBy:      "plan_finished_at asc, FIELD(priority, 'URGENT', 'HIGH', 'NORMAL', 'LOW')",
+			Priority:     IssuePriorities,
+			Type:         IssueTypes,
+			Asc:          true,
+		},
+	}
+	issues, total, err := w.issueSvc.Paging(issueReq)
+	if err != nil {
+		return nil, err
+	}
+	issueItem.TotalIssueNum = int(total)
+	issueItem.IssueList = issues
+	issueItem.ProjectDTO = pro
+	return &issueItem, nil
+}
+
+// SetSpecialIssueNum concurrent query different expire issue num
+func (w *Workbench) SetSpecialIssueNum(req apistructs.IssuePagingRequest, items []*apistructs.WorkbenchProjectItem) error {
+	if len(items) == 0 {
+		return nil
+	}
+	var projectIDS []uint64
+	for _, item := range items {
+		projectIDS = append(projectIDS, item.ProjectDTO.ID)
+	}
```

Comment:
if items only to used to retrieve projectIDs, you should let projectIDs as func param directly. you just wasted query too many sql to get the items, and I commented in another code block.

Reply:
e.workBench.GetUndoneProjectItem get issue table by issue size, not only issue num, it`s hard to group by project id list.